app/src/gecko/java/org/mozilla/focus/web/GeckoViewPermissionsDelegate.kt

```diff
@@ -0,0 +1,115 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.focus.web
+
+import android.app.Activity
+import android.content.Context
+import android.content.pm.PackageManager
+import android.net.Uri
+import android.os.Build
+import android.util.Log
+
+import org.mozilla.focus.R
+import org.mozilla.focus.permissions.IPermissionsDelegate
+import org.mozilla.gecko.GeckoSession
+
+import java.util.Locale
+
+class GeckoViewPermissionsDelegate(private val context: Context,
+                                   private val geckoSession: GeckoSession) :
+        GeckoSession.PermissionDelegate, IPermissionsDelegate {
+    private var callback: GeckoSession.PermissionDelegate.Callback? = null
+
+    override fun onRequestPermissionsResult(permissions: Array<String>, grantResults: IntArray) {
+        if (callback == null) {
+            return
+        }
+
+        val cb = callback
+        callback = null
+        for (result in grantResults) {
+            if (result != PackageManager.PERMISSION_GRANTED) {
+                // At least one permission was not granted.
+                cb!!.reject()
+                return
+            }
+        }
+        cb!!.grant()
+    }
+
+    override fun requestAndroidPermissions(session: GeckoSession, permissions: Array<String>,
+                                           callback: GeckoSession.PermissionDelegate.Callback) {
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
+            // requestPermissions was introduced in API 23.
+            callback.grant()
+        } else {
+            this.callback = callback
+            (context as Activity).requestPermissions(permissions, androidPermissionRequestCode)
+        }
+    }
+
+    override fun requestContentPermission(session: GeckoSession, uri: String,
+                                          type: String, access: String,
+                                          callback: GeckoSession.PermissionDelegate.Callback) {
+        val resId = when (type) {
+            "geolocation" -> R.string.request_geolocation
+            "desktop-notification" -> R.string.request_notification
+            else -> {
+                Log.w(LOGTAG, "Unknown permission: " + type)
+                callback.reject()
+                return
+            }
+        }
+
+        val title = context.getString(resId, Uri.parse(uri).authority)
+        val prompt = geckoSession.promptDelegate as GeckoViewPrompt
+        prompt.promptForPermission(session, title, callback)
+    }
```

Comment:
What permissions do we need this for? I think we shouldn't introduce location/camera/audio as new features at this time - unless we absolutely need to.

Reply:
With Android WebView, we can disable these permissions so that the WebView cannot even request them (https://github.com/mozilla-mobile/focus-android/blob/master/app/src/webview/java/org/mozilla/focus/web/WebViewProvider.java#L119), but with GeckoView, we can't disable these currently and if we do not implement the permissions, then the user will just see the browser as not working - see #2128.